MPF/MySTT/rasta-v2_6_1/rasta-v2_6_1/rasta.1
.\" A few notes:
.\" i) Don't add extra blank lines - these appear in the output and
.\"    leaves too much space when display on the screen
.\"
.\" This title line needs to be changed to contain your program name,
.\" but the date will be changed automatically by RCS.
.TH rasta 1 "$Date: 2001/04/13 11:00:15 $" ICSI
.\" Don't do anything clever in the section below - it messes up the
.\" whatis database that's accessed using "man -k"
.SH NAME
rasta \- computes plp or rasta-plp features.
.SH SYNOPSIS
.B rasta
.RB "[" options "]"
.SH DESCRIPTION
.I rasta
computes variants of the Perceptual Linear Prediction (PLP)
analysis, including several forms of RASTA processing. It reads
one of several possible input files containing audio waveforms,
and outputs one of several formats of feature vectors, one vector per
frame.  At least one option (described below) is required.
.SH OPTIONS
.TP 1i			\" Indent the paragraph
.B \-O " [FALSE] "
for online processing 
.TP
.B \-d " [FALSE]"
for debug output 
.TP
.B \-M " [FALSE]"
adds a small constant to the power spectrum equivalent to one bit of random
noise on the input.  This helps to avoid the numerical problems that occur
when there are long stretches of zeros in the data.
.TP
.B \-F " [FALSE]"
for highpass filter on input (for dc offset) 
.TP
.B \-a " [FALSE]"
for input ascii file.
Default input is binary shorts
.TP
.B \-A " [FALSE]"
for ascii output file. 
Default output is binary floats.
.TP
.B \-e " [FALSE]"
for ESPS format input files;
overrules other input options.
.TP
.B \-E " [FALSE]"
for ESPS format output files;
overrules other output options.
.TP
.B \-b " [FALSE]"
for MAT format input files;
overrules other input options.
.TP
.B \-B " [FALSE]"
for MAT format output files;
overrules other output options.
.TP
.B \-z " [FALSE]"
for SPHERE format input files;
overrules other input options.
.TP
.B \-k " [FALSE]"
for Abbot I/O.
.TP
.B \-X " [FALSE]"
for HTK "USER FTR" format output (12 byte header, type=9).
.TP
.B \-T " [FALSE]"
Input is little-endian.
This option makes sense only with raw binary shorts as input.
.TP
.B \-U " [FALSE]"
Output is little-endian.
This option makes sense only with raw binary floats as output.
.TP
.B \-y " [FALSE]"
Pad input to produce frames where the ((n + 0.5) * M)-th point is
centered in the n-th frame.  M is the number of points per step.  A
total of L/M frames will be produced, where L is the overall signal
length.  There must be at least one window's worth of input data to
produce output.  This option works for both normal and online (-O
option) input.
.TP
.B \-L " [FALSE]"
for log rasta 
.TP
.B \-J " [FALSE]"
for JAH rasta 
.TP
.B \-C " [FALSE]"
for constant JAH.  This option should generally be used during
recognizer training.
Default is adapting JAH according to noisepower.
Only use option -C when -J is used
.TP
.B \-R " [FALSE]"
for getting critical band values as output 
instead of cepstral coefficients.
This is useful for users who would like to find
their own spectral mapping coefficients.
This overrules the -P option.
NOTE: Formerly, -R always generated ascii output
(i.e. implied -A).  As of version 2.4.2, this 
is no longer true.  Be sure to specify -A if 
you want ascii output.
.TP
.B \-P " [FALSE]"
for getting cube root compressed and equalized
critical band values as output instead of cepstral coefficients.
NOTE: Formerly, -P always generated ascii output
(i.e. implied -A).  As of version 2.4.2, this 
is no longer true.  Be sure to specify -A if 
you want ascii output.
.TP
.B \-g " [compute gain]"
don't compute gain.
.TP
.B \-h " [FALSE]"
use stored noise level estimation and RASTA filter
history for initialization (if history file is
available, otherwise use normal initialization).
.TP
.BI \-i " input file" " [stdin]"
.TP
.BI \-o " output file" " [stdout]"
.TP
.BI \-I " list file"
This option names a file containing pairs of filenames, where the 
first of each pair is an input sound file, and the second is the 
desired corresponding output file.  Feature processing is applied to 
each input file in turn, writing a number of output files (HTK 
style).  
.TP
.BI \-f " mapping file"
JAH Rasta mapping coefficients input text file; specifies
a multi-linear regression to reduce the variance introduced by
using different J values.  There is no default, because you must
create your own mapping file.  See the examples section for the reason why.
.TP
.BI \-r " rasta mixing coefficient" " [1.0]"
for partially rasta, partially plp
e.g. -r 1.0 => no mixing, -r 0.8 => 80% rasta 20% PLP
.TP
.BI \-w " analysis window size" " [20]"
(in milliseconds)
.TP
.BI \-W " windowing constant" " [0.540000]"
.TP
.BI \-s " window step size" " [10]"
(in milliseconds)
.TP
.BI \-S " Sampling frequency" " [8000]"
(in Hertz)
.TP
.BI \-Y " Nyquist frequency" " [sampling frequency/2]"
If this is specified as less than half the sampling frequency, the 
corresponding bins from the top of the initial Fourier transform 
are discarded, performing an effective downsampling of the data.  
This doesn't exactly match actual resampling, but it's within a few 
percent.
.TP
.BI \-l " liftering exponent" " [0.600000]"
.TP
.BI \-p " pole value" " [0.940000]"
The magnitude of the integrator pole on the real axis.  This option 
is redundant with the lower cutoff frequency specified with -x; only 
one should be present.  When the 
default framerate of 10 ms is used, the default lower cutoff frequency 
gives the default pole value.  For other framerates, it is the 
lower cutoff frequency which takes precedence.  If -p is specified, 
however, it will take precedence.
.TP
.BI \-x " lower cutoff 3dB frequency" " [0.955504 Hz]"
The rasta filter includes an integrating pole close to the zero at d.c.; 
as this pole gets closer to the zero, the sharpness of the notch at 0 Hz 
increases.  To ensure that this filter performs approximately the same 
role across a range of frame rates, the pole magnitude is, by default, 
calculated to have a -3dB point at this frequency, and hence its 
value will also depend on the frame rate in use.  This option is therefore 
incompatible with -p which specifies the pole magnitude directly; they 
cannot both be used.  
.TP
.BI \-u " upper cutoff-zero frequency" " [29.021531 Hz]"
The rasta filter effects smoothing along time with its low-pass 
characteristic, which is manifested in zeros at sr/2 and at an 
intermediate frequency.  For the classic, straight-line rasta FIR 
response, this zero falls close to 29 Hz at the default frame rate 
of 100 Hz; for other frame rates, the FIR filter coefficients 
are adjusted to keep the zero at this frequency (or any other frequency 
below half the frame rate, as specified with this option).
.TP
.BI \-m " model order" " [8]"
.TP
.BI \-n " number of output parameters" 
where default is model order plus 1 (log gain) 
and log gain is given first
.TP
.BI \-c " number of crit band filters"
where default depends on sampling freq, but is 17 for 8000 Hz
.TP
.BI \-v " use triangular filters"
Use triangular auditory filters instead of the default trapezoidal filters.
.TP
.BI \-V " frequency axis warp" "[1.0]"
A value other than 1.0 causes a linear stretch (> 1.0) or squeeze (< 1.0)
of the linear frequency axis upon which the various auditory spectra 
are based.  This can be used to tune features for variations in 
vocal tract length.  See Lee & Rose, ICASSP-96.
.TP
.BI \-q " deltas order" " [0] "
The calculated features will be augmented by their derivatives (deltas)
up to the specified order (i.e. 1 for slope, 2 for slope and curvature).  
This will duplicate the feature vector size requested by -n i.e. 
-n 13 -q 2 generates a 39-element feature vector.
.TP
.BI \-Q " delta calculation window" " [9] "
The number of successive frames used in calculating the slope and 
curvature (if requested) of the features.  Derivatives are calculated 
by convolving with a simple linear-slope filter (the convolution of 
two such filters for double derivatives); this parameter controls its 
length.
.TP
.BI \-K " STRUT compatibility mode"
Make a collection of changes for compatibility with the STRUT version of 
Rasta.
Specifically, calculate a raw log energy term as the first element 
of cepstral outputs (in addition to c0), and use a double-delta 
calculation kernel that is 2 points larger than the delta window 
(for -q 2).  However, only a 7 point double-delta window is known 
for STRUT mode, so -K -q 2 must have -Q 5.  Finally, the output vector 
is sub-selected to drop the first element (the absolute value of the 
new energy term) and the last <ncep> values (the highest order deltas 
of everything 
.I except 
the new energy term).  
Note that the 
energy term is included in the -n count, so for classsic 
26 element STRUT Rasta features, you need -m 10 -n 13 -q 2 -Q 5 -K .
In this latest version, you can feed these directly into the forward pass; 
there is no need for an additional feacat filtering stage. (See BUGS).
.TP
.BI \-N " numerator rastafilt file"
unimplemented in release version 2.2
.TP
.BI \-D " denominator rastafilt file"
unimplemented in release version 2.2
.TP
.BI \-j " Constant J" " [1.000000e-06] "
depends on the noise level, ( smaller for more noisy speech).
Only use option -j when -C is used.
.TP
.BI \-H " History filename" " [history.out] "
File containing noise level and filter history for initialization.
.SH EXAMPLES
To compute log RASTA-PLP (which is optimal for the case of convolutional
error, as when the speech input has been run through a linear
time-invariant filter), with ascii input and ESPS-compatible output:

	rasta -a -E -L < infile.asc > esps_outfile

This will use all the defaults, including an assumption of 8 kHz
sampling rate for the input data. Note that the use of ESPS requires
making rasta using the ESPS-based Makefile, and also requires an
ESPS license. It will compute 9 log RASTA cepstral coefficients (8th
order, including a gain term).


To compute 12th order J-RASTA-PLP (which handles additive noise in addition
to the convolutional), using automatic noise estimation, and with
ascii input and output and a 16 kHz sampling rate:

	rasta -a -A -i speechin.asc -o speechout.asc -J -S 16000 -n 12 -f map.dat

The -f option specifies a mapping file that is used to reduce the
J-dependent variability of the rasta output by mapping spectra to a reference
J value that corresponds to clean speech.  In previous versions of this
program, there was a default value for -f, and we provided a mapping file
with the RASTA distribution.  There is no longer a default and we no longer
distribute a mapping file because the mapping file is highly dependent on the
speech being recognized.  It is absolutely vital that you derive a mapping
file from some subset of your own training data.
.BR create_mapping(1)
provides a simple mechanism for creating mapping files.

To compute PLP without any RASTA filtering, using binary shorts
at the input and binary floats at the output (and all the other
defaults):

	rasta < speechin > speechout
.SH AUTHOR
Nelson Morgan (original version, after a Fortran PLP by Hynek Hermansky)
\<morgan@icsi.berkeley.edu\>
.br \" How to break lines between authors
Later additions - Grace Tong, Chris Ehrlicher, Brian Kingsbury
\<bedk@icsi.berkeley.edu\>, Dan Ellis \<dpwe@icsi.berkeley.edu\>
.br
HTK and list file contributions by Sacha Krstulovic \<sacha@idiap.ch\>.
.SH SEE ALSO
.BR "Hermansky, H., and Morgan, N., ``Rasta Processing of Speech,''"
.ul
IEEE Transactions on Speech and Audio Processing,
special issue on Robust Speech Recognition, vol.2 no. 4, pp. 578-589,
Oct., 1994

.BR create_mapping(1)
.SH BUGS
The noise estimation procedure used in adaptive J-RASTA processing (-J
option on, -C option off) assumes that the first 100 ms. of a signal
does not contain any speech.  If this assumption is violated, then it
produces poor estimates of the noise level early in the signal.  This
bug may cause ridiculously large energy values for early frames when
the usual cepstral outputs are produced, or ridiculously large outputs
if the -P or -R option is used.  Use of a history file reduces the bad
effects of insufficient leading non-speech, but does not eliminate
them.  In general, try to make sure you have at least 100 ms. of
non-speech leader in all inputs to rasta when using the -J option.
.P
We use the soon-to-be-obsolete getopt(3) call to do command-line 
parsing.  Under gcc 2.7.1/SunOS 4.1.3, this appears to be buggy, 
so that failing to specify a required argument for a flag causes a 
core dump rather than an error message.
.P
As we run out of letters of the alphabet, command-line options are
becoming increasingly non-mnemonic.
.P
HTK output (-X) and input/output file list file (-I) are recent and
little tested.
.P
The "Strut compatibility mode" is a very evil and ugly hack.  The correct 
way to support this would be to expand the option set orthogonally so these 
options could be specified as one possible condition.  But (a) the option 
set is already a non-orthogonal evil mess (for orthogonality, use 
.MS feacat 1 )
and (b) we don't have enough options left to do it anyway!
.P
Doesn't yet implement the -N or -D switches to specify the rasta filter.
It also doesn't eliminate all problems
in speech recognition yet.


MPF/MySTT/rasta-v2_6_1/rasta.c
/************************************************************************
 *                                                                       *
 *               ROUTINES IN THIS FILE:                                  *
 *                                                                       *
 *                      main(): main calling routine                     *
 *                                                                       *
 ************************************************************************/

/***********************************************************************

 (C) 1995 US West and International Computer Science Institute,
     All rights reserved
 U.S. Patent Numbers 5,450,522 and 5,537,647

 Embodiments of this technology are covered by U.S. Patent
 Nos. 5,450,522 and 5,537,647.  A limited license for internal
 research and development use only is hereby granted.  All other
 rights, including all commercial exploitation, are retained unless a
 license has been granted by either US West or International Computer
 Science Institute.

***********************************************************************/

/********************************************************************
	Scope of Algorithm


The actual processing is split up into a few pieces:

        1) power spectral analysis
        2) auditory spectrum computation
        3) compression (possibly adaptive)
        4) temporal filtering of each channel trajectory
        5) expansion (also possibly adaptive)
        6) postprocessing (e.g., preemphasis, loudness compression)
        7) autoregressive all-pole modeling (cepstral coefficients)

and i/o can either be ascii, binary (shorts on the input and floats
on the output), or standard ESPS files.

Since the lowest frequency and highest frequency bands extend
into forbidden territory (negative or greater-than-Nyquist
frequencies), they are ignored for most of the analysis. This
is done by computing a variable called first_good (which for
1 bark spacing is 1) and generally ignoring the first and last
``first_good'' channels. Just prior to the all-pole modeling,
the values from the good channels are copied over to the questionable
ones. (Note that ``first_good'' is available to most routines
via a pointer to a general parameter structure that has such
useful things as the number of filters, the analysis stepsize
in msec, the sampling rate, etc. . See rasta.h for
a definition of this structure).


This program (Rasta 2.0) implements the May 1994 version of RASTA-PLP
analysis. It incorporates several primary pieces:

1) PLP Analysis - as described in Hermansky's 1990 JASA paper,
the basic form of spectral analysis is Perceptual Linear Prediction,
or PLP. This computes the cepstral parameters of an all-pole model
of an auditory spectrum, which is a power spectrum that
has been frequency warped to the bark scale, smoothed by
an asymmetric critical-band trapezoid function,
down-sampled into approximately 1 Bark intervals,
cube root compressed for an intensity-loudness transformation, 
and weighted by a fixed equal loudness curve. 

2) RASTA filtering - as described in several Hermansky and Morgan
papers, the basic idea here is to bandpass filter the trajectories
of the spectral parameters. In the case of RASTA-PLP, this filtering
is done on a nonlinear transformation of an auditory-like spectrum, 
prior to the autoregressive modeling in PLP.

3) J-processing - For RASTA, the bandpass filtering is done in the 
log domain. An alternative is to use the J-family
of log-like curves

	y = log(1 + Jx)

where J is a constant that can appears to be optimally set when
it is inversely proportional to the noise power, (currently typically
1/3 of the inverse noise), x is the
critical band value, and y is the non-linearly transformed critical 
band value.
Rather than do the true inverse, which would be

	x = (exp(y) - 1)/J

and could get negative values, we use

	x' = (exp(y))/J

This prevents negative values, and in doing so effectively adds noise
floor by adding 1/J to the true inverse. 

One way of doing J-processing is to pick one constant J value and enter
this value at the command line. This J value should be dependent on the
SNR of the speech. We also may want to estimate the noise level for 
adaptive settings of the J-parameter during the utterance. 
Both methods of picking J should be handled with care. For the first
case, see the README file for a discussion of the perils of using even
a default J if it is too far from what you really need;
if the application situation is relatively fixed, you are better off making
an off-line noise measurement to get a good J value; in any event some
experimentation will soon show the proper constants involved for a problem.
For the second case, noise level is estimated for adaptive settings of the
J-parameter during the utterance. This should be done with care as well, 
as the use of a time-varying J brings in a new complication that you must
consider in the training and recognition, since changing J's over a time
series introduces a new source of variability in the analysis that must
be accounted for. The different J values, as required by differing noise
conditions, generates different spectral shapes and dynamics of the spectra.
This means that the training system must contend with a new source of 
variability due to the change in processing strategy that is adaptively
determined from the data. One approach to handle this variability is by
doing spectral mapping. In the current version, Spectral mapping is 
performed whenever J-Rasta processing is used with adaptive Js.

Spectral mapping - transform the spectrum processed with a J-value  
                   corresponding to noisy speech to a spectrum processed
                   with a J value for clean speech. In other words, we
                   find a mapping between log(1+xJ) and log(1+xJref) 
                   where Jref is the reference J, i.e. J value for clean speech.
                   For this approach, we have used a multiple regression 
                   within each critical band. In principle, this solution
                   reduces the variability due to the choice of J, and so
                   minimizes the effect on the training process.
       

How this works is:

1) Training of the recognizer:
   -- Train the recognizer with clean speech processed with J = 1.0e-6, a 
      suitable J value for clean speech.

2) Finding the relationship between spectrum corresponding to different Jah 
   values to the spectrum corresponding to J = 1.0e-6
   -- For each of the Jahs in the set {3.16e-7, 1.0e-7, 3.16e-8, 1.0e-8, 
      3.16e-9, 1.0e-9}, find a mapping relationship of the corresponding 
      bandpass filtered spectrum to the spectrum corresponding to J =
      1.0e-6. In other words, find a set of mapping coefficients for each
      Jah to 1.0e-6. The mapping method will be discussed later.

3) Extracting the speech features for the testing speech data
   -- obtain the critical band values as in PLP
   -- estimate the noise energy and thus the Jah value. Call this Jah value 
      J(orig).
   -- Pick a Jah from the set {3.16e-7, 1.0e-7, 3.16e-8, 1.0e-8,3.16e-9, 1.0e-9}
      that is closest to J(orig) and call this J(quant).
   -- perform the non-linear transformation of the spectrum using 
      log (1+J(quant)* X).
   -- bandpass filter the transformed critical band values.
   -- use the set of mapping coefficients for J(quant) to do the spectral 
      mapping or spectral transformation.
   -- preemphasize via the equal-loudness curve and then perform amplitude
      compression using the intensity-loudness power law as in PLP
   -- take the inverse of the non-linear function. 
   -- compute the cepstral parameters for the AR model.   
    
    
 
How regression coefficients are computed in our experiment:

    In order to map critical band values(after bandpass filtering) processed
    with different J values to those processed with J = 1.0e-6, J-Rasta 
    filtered critical band outputs from 10 speakers(excluded from the training
    and testing sets) are used to train 
    multiple regression models.  
    For example, for mapping from J= J(quant) to J = 1.0e-6, the regression
    equation can be written as:

    Yi = B2i* X2 + B3i* X3 + ... + B16i * X16 + B17i       (**)

where Yi = i th bandpass filtered critical band processed with J=1.0e-6
           i = 2, .. 16
      X2, X3, ... X16    2rd to 16th bandpass filtered critical band values
                         processed with J = J(quant), where
                         J(quant) is in the set 
                         {3.16e-7, 1.0e-7, 3.16e-8, 1.0e-8,3.16e-9, 1.0e-9} 
      B2i, B3i ... B17i     are the 16  mapping coefficients 

      
    For equation (**), we have made the assumption that the sampling frequency
    is 8kHz and the number of critical bands is seventeen. The first and 
    the last bands extend into forbidden territory -- negative or greater
    than Nyquist frequencies. Thus the the two bands are ignored for most
    of the analysis. Their values are made equal to the adjacent band's just
    before the autoregressive all-pole modeling. This is why we only make
    Yi dependent on bandpass filtered critical bands X2,X3,... X16, altogether
    fifteen critical bands. 

    The default mapping coefficients sets is stored in map_weights.dat. 
    This is suitable for s.f. 8kHz, 17 critical bands. For users who have 
    a different setup, they may want to find their own mapping coefficients
    set. This could be done by using the command options -R and -f. Command 
    -R allows you to get bandpass filtered critical band values as output
    instead of cepstral coefficients. 
    These outputs could be used as regression data. A simple multiple
    regression routine can be used to generate the mapping coefficients
    from these regression data. These mapping coefficients can be stored
    in a file. Command -f allows you to use this file to replace the
    default file map_weights.dat. The format of this file is:

     beginning of file 
    <Total number of Jahs, for the example shown above, it is [7] > 
    <# of critical bands, for the setup for 8kHz, this is [15]>
    <# of mapping coefficients/band, for the setup for 8kHz, this is [16]>
    
    <The J for clean speech, [1.0e-6]>
   
    <mapping coefficients for Y2, [B22, B32, B42,...]>
    <mapping coefficients for Y3, [B23, B33, B43 ...]>
        |
        |
        |
        V
    <mapping coefficients for Y16>
 
    
    <The second largest Jah besides 1e-6, [3.16e-7]>
        
        |
        |      mapping coefficients
        |
        |
        V 
  
    <The third largest Jah besides 1e-6, [1.0e-7]>
       
        |
        |      mapping coefficients
        |
        |
        V 
  
     .
     .
     .
     .    
     end of file



*********************************************************************/
#include <stdio.h>
#include <math.h>
#include "rasta.h"

/******************************************************/

static void reset_state(struct param *runparam, struct fhistory *history) {
    /* reset state in history structure, rasta filter, and
       noise estimation. */

    history->normInit = TRUE;
    if (runparam->history == TRUE) {
	load_history(history, runparam);
    }

    history->eos = TRUE;	/* causes rasta_filt to reset */
    rasta_filt(history, runparam, (struct fvec*) NULL);

    if ((runparam->jrasta == TRUE ) &&
	(runparam->cJah == FALSE)) { 
	comp_noisepower(history, runparam, (struct fvec*) NULL);
    }

    history->eos = FALSE;
}


int
main(int argc,char** argv)
{
  /* function prototypes for local calls */
  void init_param(struct fvec*, struct param*);
  struct fvec* get_data(struct param*);
  struct fvec* rastaplp(struct fhistory*,
                        struct param*,
                        struct fvec*);
  struct fvec* fill_frame(struct fvec*,
                          struct fhistory*,
                          struct param*,
                          int);

  /*  	Variables and data structures	*/
  struct param runparam;
  struct fhistory history;
  struct fvec *speech, *cepstrum, *frame, *plusdeltas;
  int nframe;
  char *funcname;
  FILE *outfp;
  FILE *lstfp;           /* SK */
  char lstbuffin[512];   /* SK */
  char lstbuffout[512];  /* SK */
  DeltaCalc *dc;

  if(argc == 1) { 
      /* print out usage if no arguments used */
      usage(argv[0]);
      exit(0);
  }

  funcname = "main";

  get_comline(&runparam, argc,argv); /* preferably list arguments in 
					command line so there is a record */

  check_args( &runparam ); /* Exits if params out of range */

  /* If the debugging is on, show the command-line argurments */

  if (runparam.debug == TRUE) {
      show_args( &runparam );
  }

  if (runparam.IOlist == TRUE) {  /* Using filelist */
      /* Open filelist */
      if ( ! ( lstfp = fopen( runparam.list_fname , "r" )) ) {
          fprintf(stderr,"Unable to open file list.\n");
          exit(-1);
      }
  }

  if (runparam.infname == NULL) {
      runparam.infname = "-";
  }
  if (runparam.outfname == NULL) {
      runparam.outfname = "-";
  }

  history.eof = FALSE;

  /* initialize delta calculation */
  dc = alloc_DeltaCalc(0 /* runparam.nout */, runparam.deltawindow, 
		       runparam.deltaorder, runparam.strut_mode);
  /* You can't actually tell the width of the output features unless 
     you examine the cbrout and comcbrout flags.  Passing 0 tells the 
     DeltaCalc code to defer deciding the feature width until the 
     first frame is seen, and take it from that! dpwe 1997aug13 */
  
  /* preset the 'speech' structure to something we'll know not to use */
  speech = (struct fvec *)NULL;

  do {    /* Filelist loop (executed just once if !IOlist) */
      
      if (runparam.IOlist == TRUE) {  /* Using filelist */
          if ( fscanf(lstfp,"%s %s",lstbuffin,lstbuffout) != 2 ) {
	      break;
	  } else { 
	      runparam.infname = lstbuffin;
	      runparam.outfname = lstbuffout;
	  }
      }

      if ( runparam.online == FALSE && runparam.abbotIO == FALSE) {
	  if (speech) { free_fvec(speech); }
	  speech = get_data(&runparam); /* allocate for and read data */
      }

      outfp = open_out( &runparam ); /* Open output file */

      init_param(speech, &runparam); /* Compute necessary parameters 
					for analysis */

      reset_state(&runparam, &history);

      /* main analysis loop */

      for (nframe = 0; nframe < runparam.nframes; ) {

	  if (runparam.online == TRUE) {
	      frame = get_online_bindata( &history, &runparam );
	  } else if (runparam.abbotIO == TRUE) {
	      frame = get_abbotdata(&history, &runparam);
	  } else {
	      frame = fill_frame( speech, &history, &runparam, nframe);
	      /* only count frames if not online - else rp.nframes == 1 */
	      nframe++;
	  }

          if (history.eos) {
	      /* terminate the loop by warping to the end condition */
	      nframe = runparam.nframes;
	  } else {
	      cepstrum = rastaplp( &history, &runparam, frame );

	      plusdeltas = DeltaCalc_appendDeltas(dc, cepstrum);
	      
	      if(plusdeltas->length) {
		  struct fvec tmpout;
		  if (runparam.strut_mode) {
		      /* implement the strut-style Cep+dE+dCep+ddE */
		      tmpout.values = plusdeltas->values + 1;
		      tmpout.length = (runparam.deltaorder)*cepstrum->length;
		  } else {
		      tmpout.values = plusdeltas->values;
		      tmpout.length = plusdeltas->length;
		  }
		  write_out( &runparam, outfp, &tmpout );
	      }
	  }
      }

      /* flush deltas state */
      plusdeltas = DeltaCalc_appendDeltas(dc, NULL);
      while(plusdeltas->length) {
	  struct fvec tmpout;
	  if (runparam.strut_mode) {
	      /* implement the strut-style Cep+dE+dCep+ddE */
	      tmpout.values = plusdeltas->values + 1;
	      tmpout.length = (runparam.deltaorder)*cepstrum->length;
	      /* luckily, cepstrum->length is still available... */
	  } else {
	      tmpout.values = plusdeltas->values;
	      tmpout.length = plusdeltas->length;
	  }
	  write_out( &runparam, outfp, &tmpout );
	  plusdeltas = DeltaCalc_appendDeltas(dc, NULL);
      }
      DeltaCalc_reset(dc);

      if (runparam.abbotIO == TRUE) {
	  abbot_write_eos(&runparam);
      }

      if (outfp != stdout) {
	  fclose(outfp);
      }

/*      fprintf(stderr, "rasta: bottom of loop: iolist=%d abbotio=%d online=%d eof=%d\n", runparam.IOlist,  runparam.abbotIO, runparam.online, history.eof);
*/
  } while ( (runparam.IOlist == TRUE && !feof(lstfp)) \
	   || (runparam.abbotIO == TRUE && !history.eof) \
           || (runparam.online == TRUE && !history.eof));

  /* fprintf(stderr, "rasta: exiting\n"); */
  /* clean up */
  free_DeltaCalc(dc);

  if (speech) {
      free(speech->values);
      free(speech);
  }

  if(runparam.history == TRUE) {
      save_history(&history, &runparam);
  }

  return(0);
}
